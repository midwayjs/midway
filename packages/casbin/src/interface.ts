import { Adapter, Watcher } from 'casbin';
import { AuthActionVerb, AuthPossession } from './constants';
import { IMidwayContainer, IMidwayContext } from '@midwayjs/core';

export interface CasbinConfigOptions {
  modelPath: string;
  policyAdapter: string | ((applicationContext: IMidwayContainer) => Promise<Adapter>) | Adapter;
  policyWatcher?: ((applicationContext: IMidwayContainer) => Promise<Watcher>) | Watcher;
  usernameFromContext: (ctx: IMidwayContext) => string;
}

export interface Permission {
  resource: string;
  action: AuthActionVerb | CustomAuthActionVerb;
  possession: AuthPossession;
  isOwn?: (ctx: IMidwayContext) => boolean;
}

export type CustomAuthActionVerb = string;
