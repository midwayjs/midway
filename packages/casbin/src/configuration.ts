import { Configuration } from '@midwayjs/core';
import { CasbinEnforcerService } from './enforcer.service';

@Configuration({
  namespace: 'casbin',
  importConfigs: [
    {
      default: {
        casbin: {},
      },
    },
  ],
})
export class CasbinConfiguration {
  async onReady(container) {
    await container.getAsync(CasbinEnforcerService);
  }
}
