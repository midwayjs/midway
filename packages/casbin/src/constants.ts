export const PERMISSIONS_METADATA_KEY = 'casbin:permission_metadata';

export enum AuthActionVerb {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
  READ = 'read',
}

export enum AuthPossession {
  ANY = 'any',
  OWN = 'own',
  OWN_ANY = 'own|any',
}

export enum AuthAction {
  CREATE_ANY = 'create:any',
  CREATE_OWN = 'create:own',

  UPDATE_ANY = 'update:any',
  UPDATE_OWN = 'update:own',

  DELETE_ANY = 'delete:any',
  DELETE_OWN = 'delete:own',

  READ_ANY = 'read:any',
  READ_OWN = 'read:own',
}
