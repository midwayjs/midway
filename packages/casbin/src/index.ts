export { CasbinConfiguration as Configuration } from './configuration';
export * from './enforcer.service';
export * from './auth.guard';
export * from './decorator';
export * from './interface';
export * from './constants';
export * from './adapter';
