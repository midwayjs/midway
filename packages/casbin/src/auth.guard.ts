import {
  Config,
  getPropertyMetadata,
  Guard,
  IGuard,
  IMidwayContext,
  Inject,
} from '@midwayjs/core';
import { CasbinEnforcerService } from './enforcer.service';
import { CasbinConfigOptions, Permission } from './interface';
import { AuthPossession, PERMISSIONS_METADATA_KEY } from './constants';

@Guard()
export class AuthGuard implements IGuard {
  @Inject()
  enforcer: CasbinEnforcerService;

  @Config('casbin')
  casbinConfig: CasbinConfigOptions;

  async canActivate(
    context: IMidwayContext,
    supplierClz,
    methodName: string
  ): Promise<boolean> {
    const permissions: Permission[] = getPropertyMetadata<Permission[]>(
      PERMISSIONS_METADATA_KEY,
      supplierClz,
      methodName
    );

    if (!permissions) {
      return true;
    }

    const username = this.casbinConfig.usernameFromContext(context);

    if (!username) {
      return false;
    }

    const hasPermission = async (
      user: string,
      permission: Permission
    ): Promise<boolean> => {
      const { possession, resource, action } = permission;
      const poss = [];

      if (possession === AuthPossession.OWN_ANY) {
        poss.push(AuthPossession.ANY, AuthPossession.OWN);
      } else {
        poss.push(possession);
      }

      return AuthGuard.asyncSome<AuthPossession>(poss, async p => {
        if (p === AuthPossession.OWN) {
          return (permission as any).isOwn(context);
        } else {
          return this.enforcer.enforce(user, resource, `${action}:${p}`);
        }
      });
    };

    return await AuthGuard.asyncEvery<Permission>(
      permissions,
      async permission => hasPermission(username, permission)
    );
  }

  static async asyncSome<T>(
    array: T[],
    callback: (value: T, index: number, a: T[]) => Promise<boolean>
  ): Promise<boolean> {
    for (let i = 0; i < array.length; i++) {
      const result = await callback(array[i], i, array);
      if (result) {
        return result;
      }
    }

    return false;
  }

  static async asyncEvery<T>(
    array: T[],
    callback: (value: T, index: number, a: T[]) => Promise<boolean>
  ): Promise<boolean> {
    for (let i = 0; i < array.length; i++) {
      const result = await callback(array[i], i, array);
      if (!result) {
        return result;
      }
    }

    return true;
  }
}
