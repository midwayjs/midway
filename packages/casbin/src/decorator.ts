import { Permission } from './interface';
import { PERMISSIONS_METADATA_KEY } from './constants';
import { IMidwayContext, savePropertyMetadata } from '@midwayjs/core';

const defaultIsOwn = (ctx: IMidwayContext): boolean => false;

/**
 * You can define multiple permissions, but only
 * when all of them satisfied, could you access the route.
 */
export function UsePermission(...permissions: Permission[]): MethodDecorator {
  return (target, propertyKey, descriptor) => {
    const perms = permissions.map(item => {
      if (!item.isOwn) {
        item.isOwn = defaultIsOwn;
      }
      return item;
    });

    savePropertyMetadata(PERMISSIONS_METADATA_KEY, perms, target, propertyKey);
  };
}
