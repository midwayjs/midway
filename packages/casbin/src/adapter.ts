import { FilteredAdapter, Helper, Model } from 'casbin';

interface Line {
  ptype: string;
  v0: string;
  v1: string;
  v2: string;
  v3: string;
  v4: string;
  v5: string;
}

export abstract class BaseAdapter<AdapterLine extends Line>
  implements FilteredAdapter
{
  private filtered = false;
  private policies: AdapterLine[];

  public isFiltered(): boolean {
    return this.filtered;
  }

  public async loadPolicy(model: Model): Promise<void> {
    const policies = await this.loadPolicyByAdapter();
    this.policies = policies;
    for (const line of policies) {
      this.loadPolicyLine(line, model);
    }
  }

  public async savePolicy(model: Model): Promise<boolean> {
    const policyRuleAST = model.model.get('p')!;
    const groupingPolicyAST = model.model.get('g')!;
    const policies: AdapterLine[] = [];

    for (const astMap of [policyRuleAST, groupingPolicyAST]) {
      for (const [ptype, ast] of astMap) {
        for (const rule of ast.policy) {
          const line = this.savePolicyLine(ptype, rule);
          policies.push(line);
        }
      }
    }

    await this.savePolicyByAdapter(policies);
    return true;
  }

  public async addPolicy(
    sec: string,
    ptype: string,
    rule: string[]
  ): Promise<void> {
    const line = this.savePolicyLine(ptype, rule);
    this.policies.push(line);
    await this.savePolicyByAdapter(this.policies);
  }

  public async removePolicy(
    sec: string,
    ptype: string,
    rule: string[]
  ): Promise<void> {
    const filteredPolicies = this.policies.filter(policy => {
      let flag = true;
      flag &&= ptype === policy.ptype;
      if (rule.length > 0) {
        flag &&= rule[0] === policy.v0;
      }
      if (rule.length > 1) {
        flag &&= rule[1] === policy.v1;
      }
      if (rule.length > 2) {
        flag &&= rule[2] === policy.v2;
      }
      if (rule.length > 3) {
        flag &&= rule[3] === policy.v3;
      }
      if (rule.length > 4) {
        flag &&= rule[4] === policy.v4;
      }
      if (rule.length > 5) {
        flag &&= rule[5] === policy.v5;
      }
      return !flag;
    });
    this.policies = filteredPolicies;
    const line = this.savePolicyLine(ptype, rule);
    await this.removePolicyByAdapter(line, filteredPolicies);
  }

  public async loadFilteredPolicy(model: Model, filter: any): Promise<void> {
    const filteredPolicies = await this.loadPolicyWithFilterByAdapter(filter);
    this.policies = filteredPolicies;
    filteredPolicies.forEach((policy: any) => {
      this.loadPolicyLine(policy, model);
    });
    this.filtered = true;
  }

  public async removeFilteredPolicy(
    sec: string,
    ptype: string,
    fieldIndex: number,
    ...fieldValues: string[]
  ): Promise<void> {
    const rule = new Array<string>(fieldIndex).fill('');
    rule.push(...fieldValues);
    const filteredPolicies = this.policies.filter(policy => {
      let flag = true;
      flag &&= ptype === policy.ptype;
      if (rule.length > 0 && rule[0]) {
        flag &&= rule[0] === policy.v0;
      }
      if (rule.length > 1 && rule[1]) {
        flag &&= rule[1] === policy.v1;
      }
      if (rule.length > 2 && rule[2]) {
        flag &&= rule[2] === policy.v2;
      }
      if (rule.length > 3 && rule[3]) {
        flag &&= rule[3] === policy.v3;
      }
      if (rule.length > 4 && rule[4]) {
        flag &&= rule[4] === policy.v4;
      }
      if (rule.length > 5 && rule[5]) {
        flag &&= rule[5] === policy.v5;
      }
      return !flag;
    });
    this.policies = filteredPolicies;

    const line = new (this.getAdapterLine())();

    line.ptype = ptype;

    if (fieldIndex <= 0 && 0 < fieldIndex + fieldValues.length) {
      line.v0 = fieldValues[0 - fieldIndex];
    }
    if (fieldIndex <= 1 && 1 < fieldIndex + fieldValues.length) {
      line.v1 = fieldValues[1 - fieldIndex];
    }
    if (fieldIndex <= 2 && 2 < fieldIndex + fieldValues.length) {
      line.v2 = fieldValues[2 - fieldIndex];
    }
    if (fieldIndex <= 3 && 3 < fieldIndex + fieldValues.length) {
      line.v3 = fieldValues[3 - fieldIndex];
    }
    if (fieldIndex <= 4 && 4 < fieldIndex + fieldValues.length) {
      line.v4 = fieldValues[4 - fieldIndex];
    }
    if (fieldIndex <= 5 && 5 < fieldIndex + fieldValues.length) {
      line.v5 = fieldValues[5 - fieldIndex];
    }

    return this.removePolicyByAdapter(line, filteredPolicies);
  }

  private loadPolicyLine(line: any, model: any) {
    const lineText =
      line.ptype +
      ', ' +
      [line.v0, line.v1, line.v2, line.v3, line.v4, line.v5]
        .filter(n => n)
        .join(', ');
    Helper.loadPolicyLine(lineText, model);
  }

  private savePolicyLine(ptype: string, rule: string[]): AdapterLine {
    const line = new (this.getAdapterLine())();

    line.ptype = ptype;
    if (rule.length > 0) {
      line.v0 = rule[0];
    }
    if (rule.length > 1) {
      line.v1 = rule[1];
    }
    if (rule.length > 2) {
      line.v2 = rule[2];
    }
    if (rule.length > 3) {
      line.v3 = rule[3];
    }
    if (rule.length > 4) {
      line.v4 = rule[4];
    }
    if (rule.length > 5) {
      line.v5 = rule[5];
    }

    return line;
  }

  protected abstract getAdapterLine(): new () => AdapterLine;
  protected abstract loadPolicyByAdapter(): Promise<AdapterLine[]>;
  protected abstract loadPolicyWithFilterByAdapter(
    filter
  ): Promise<AdapterLine[]>;
  protected abstract savePolicyByAdapter(
    policies: AdapterLine[]
  ): Promise<void>;
  protected abstract removePolicyByAdapter(
    removePolicy: AdapterLine,
    newestPolicies?: AdapterLine[]
  ): Promise<void>;
}
