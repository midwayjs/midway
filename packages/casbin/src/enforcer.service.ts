import { newEnforcer, Enforcer } from 'casbin';
import {
  ApplicationContext,
  Config,
  delegateTargetAllPrototypeMethod,
  IMidwayContainer,
  Init,
  Provide,
  Scope,
  ScopeEnum,
} from '@midwayjs/core';
import { CasbinConfigOptions } from './interface';

@Provide()
@Scope(ScopeEnum.Singleton)
export class CasbinEnforcerService {
  private instance: Enforcer;

  @Config('casbin')
  protected casbinConfig: CasbinConfigOptions;

  @ApplicationContext()
  applicationContext: IMidwayContainer;

  @Init()
  protected async init() {
    if (typeof this.casbinConfig.policyAdapter === 'function') {
      this.casbinConfig.policyAdapter = await this.casbinConfig.policyAdapter(
        this.applicationContext
      );
    }

    if (typeof this.casbinConfig.policyWatcher === 'function') {
      this.casbinConfig.policyWatcher = await this.casbinConfig.policyWatcher(
        this.applicationContext
      );
    }

    this.instance = await newEnforcer(
      this.casbinConfig.modelPath,
      this.casbinConfig.policyAdapter
    );

    if (this.casbinConfig.policyWatcher) {
      this.instance.setWatcher(this.casbinConfig.policyWatcher);
    }
  }

  public getEnforcer(): Enforcer {
    return this.instance;
  }
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface CasbinEnforcerService extends Enforcer {
  // empty
}

delegateTargetAllPrototypeMethod(CasbinEnforcerService, Enforcer);
