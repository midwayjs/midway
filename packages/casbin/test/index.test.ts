import { createApp, createHttpRequest } from '@midwayjs/mock';
import { CasbinEnforcerService } from '../src';
import { IMidwayApplication } from '@midwayjs/core';

describe('/test/index.test.ts', () => {
  it('should test casbin api', async () => {
    const app = await createApp('base-app') as IMidwayApplication;
    const enforcerService = await app.getApplicationContext().getAsync<CasbinEnforcerService>(CasbinEnforcerService);
    const policy = await enforcerService.getPolicy();
    expect(policy).toMatchSnapshot();
  });

  it('should test decorator', async () => {
    const app = await createApp('base-app-decorator') as IMidwayApplication;
    const response = await createHttpRequest(app).get('/?username=bob');
    expect(response.status).toEqual(403);

    const response1 = await createHttpRequest(app).get('/?username=tom');
    expect(response1.status).toEqual(200);
  });
});
