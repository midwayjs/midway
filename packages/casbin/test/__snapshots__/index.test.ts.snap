// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/index.test.ts should test casbin api 1`] = `
[
  [
    "alice",
    "data1",
    "read",
  ],
  [
    "bob",
    "data2",
    "write",
  ],
]
`;
